asm/ddot_test.go

```diff
@@ -0,0 +1,172 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package asm
+
+import (
+	"math"
+	"testing"
+)
+
+func TestDdotUnitary(t *testing.T) {
+	for i, test := range []struct {
+		xData []float64
+		yData []float64
+
+		want float64
+	}{
+		{
+			xData: []float64{2},
+			yData: []float64{-3},
+			want:  -6,
+		},
+		{
+			xData: []float64{0, 0, 1, 1, 2, -3, -4},
+			yData: []float64{0, 1, 0, 3, -4, 5, -6},
+			want:  4,
+		},
+	} {
+		x, xFront, xBack := newGuardedVector(test.xData, 1)
+		y, yFront, yBack := newGuardedVector(test.yData, 1)
+		got := DdotUnitary(x, y)
+
+		if !allNaN(xFront) || !allNaN(xBack) {
+			t.Errorf("test %v: out-of-bounds write to first argument", i)
```

Comment:
When you fail here, actually print the guards so it can be seen without modifying the tests - also cases below.